### PR TITLE
Add support for Singularity

### PR DESCRIPTION
**Pull request by natacha-beck**

### CODE REVIEW

File: BrainPortal/app/controllers/tool_configs_controller.rb [hunk 1]
```diff
@@ -167,7 +167,7 @@ def update #:nodoc:
     form_tool_config.bourreau_id = @tool_config.bourreau_id
 
     # Update everything else
-    [ :version_name, :description, :script_prologue, :group_id, :ncpus, :docker_image,
+    [ :version_name, :description, :script_prologue, :group_id, :ncpus, :docker_image, :singularity_image_userfile_id,
```
Comment (crocodoyle) [inline BrainPortal/app/controllers/tool_configs_controller.rb]: why do we have a different convention for `:docker_image` and `:signularity_image_userfile_id`?

File: BrainPortal/app/views/tool_configs/_form_fields.html.erb [hunk 2]
```diff
@@ -108,6 +108,18 @@
     </p>
   </span>
 
+    <span title="Id of the Singularity image">
```
Comment (crocodoyle) [inline BrainPortal/app/views/tool_configs/_form_fields.html.erb]: ID

File: BrainPortal/app/views/tool_configs/_form_fields.html.erb [hunk 3]
```diff
@@ -108,6 +108,18 @@
...
+    <p><%= f.label :singularity_image_userfile_id, "ID of the Singularity image:" %><br/>
+      <%= f.text_field :singularity_image_userfile_id %> <%= link_to_userfile_if_accessible(@tool_config.singularity_image) %><br>
+      <div class="field_explanation">
+        The ID number of the Singularity image in which the tool is installed.
+        This ID refers to userfile register in CBRAIN.
```
Comment (crocodoyle) [inline BrainPortal/app/views/tool_configs/_form_fields.html.erb]: *registered

File: BrainPortal/app/controllers/tool_configs_controller.rb (hunk 1 above)
Comment (natacha-beck) [inline BrainPortal/app/controllers/tool_configs_controller.rb]: It is normal. The support for Docker pull a `docker_image` from dockerhub, and the support for Singularity  use a `singularity_image_userfile_id` registered in CBRAIN.

**Comment by natacha-beck** [comment]
@prioux, I just fixed Andrew comment´s. If you want to take a look.

File: BrainPortal/app/models/cluster_task.rb [hunk 4]
```diff
@@ -1969,6 +1977,71 @@ def docker_commands
...
+    singularity_image = self.tool_config.singularity_image
+    singularity_image.sync_to_cache
+    cachename         = singularity_image.cache_full_path
+    safe_symlink(cachename,singularity_image.name)
+
+    # Used to set dp_cache for singularity
+    basename_dp_cache         = ".singularity_dp_cache"
+    singularity_dp_cache_path = "#{CBRAIN::Rails_UserHome}/#{basename_dp_cache}"
+
+    begin
```
Comment (prioux) [inline BrainPortal/app/models/cluster_task.rb]: Give an explanation of what that code block does, and why.

File: BrainPortal/app/models/cluster_task.rb [hunk 5]
```diff
@@ -1969,6 +1977,71 @@ def docker_commands
...
+  end
+
+  # Return the 'singularity' command to be used for the task; this is fetched
+  # from the Bourreau's own attribute. Default: "singularity".
+  def singularity_executable_name
+    return self.bourreau.singularity_executable_name.presence || "singularity"
+  end
+
+  # Return whether the Bourreau of the task has singularity present or not
+  def singularity_present?
```
Comment (prioux) [inline BrainPortal/app/models/cluster_task.rb]: This is not used at all?

Comment (natacha-beck) [inline BrainPortal/app/models/cluster_task.rb]: It is used here: https://github.com/natacha-beck/cbrain/blob/60ef22e463aa1f3d2baca3f7c22ef4907aab2f07/BrainPortal/config/console_rc/lib/pretty_view.rb#L262

Comment (prioux) [inline BrainPortal/app/models/cluster_task.rb]: No it isn't. The method in pretty_view.rc is invoked on a Bourreau. Here I'm talking about the CbrainTask method. It isn't used anywhere.

**Pull request merged**